toolchain/parse/handle_match.cpp

```diff
@@ -0,0 +1,232 @@
+                        "Expected `{{` after `=>`.");
+      context.emitter().Emit(*context.position(), ExpectedMatchCaseBlock);
+    }
+
+    context.AddNode(starter, *context.position(), state.subtree_start, true);
+
+    context.AddNode(complete, *context.position(), state.subtree_start, true);
+
+    context.SkipPastLikelyEnd(*context.position());
+    return;
+  }
+
+  context.AddNode(starter, context.Consume(), state.subtree_start,
+                  state.has_error);
+  context.PushState(state, finish);
+  context.PushState(State::StatementScopeLoop);
+}
+}  // namespace
+
+auto HandleMatchIntroducer(Context& context) -> void {
+  auto state = context.PopState();
+  context.AddLeafNode(NodeKind::Placeholder, *context.position());
+  context.PushState(state, State::MatchConditionFinish);
+  context.PushState(State::ParenConditionAsMatch);
+  context.ConsumeAndDiscard();
+}
+
+auto HandleMatchConditionFinish(Context& context) -> void {
+  auto state = context.PopState();
+  context.ReplacePlaceholderNode(state.subtree_start, NodeKind::MatchIntroducer,
+                                 state.token);
+
+  if (!context.PositionIs(Lex::TokenKind::OpenCurlyBrace)) {
+    if (!state.has_error) {
+      CARBON_DIAGNOSTIC(ExpectedMatchCasesBlock, Error,
+                        "Expected `{{` starting block with cases.");
+      context.emitter().Emit(*context.position(), ExpectedMatchCasesBlock);
+    }
+
+    context.AddNode(NodeKind::MatchStatementStart, *context.position(),
+                    state.subtree_start, true);
+
+    context.AddNode(NodeKind::MatchStatement, *context.position(),
+                    state.subtree_start, true);
+
+    context.SkipPastLikelyEnd(*context.position());
+    return;
+  }
+
+  context.AddNode(NodeKind::MatchStatementStart, context.Consume(),
+                  state.subtree_start, state.has_error);
+
+  state.has_error = false;
+  if (context.PositionIs(Lex::TokenKind::CloseCurlyBrace)) {
+    CARBON_DIAGNOSTIC(ExpectedMatchCases, Error, "Expected cases.");
+    context.emitter().Emit(*context.position(), ExpectedMatchCases);
+    state.has_error = true;
+  }
+
+  context.PushState(state, State::MatchStatementFinish);
+  context.PushState(State::MatchCaseLoop);
+}
+
+auto HandleMatchCaseLoop(Context& context) -> void {
+  context.PopAndDiscardState();
+
+  if (context.PositionIs(Lex::TokenKind::Case)) {
+    context.PushState(State::MatchCaseLoop);
+    context.PushState(State::MatchCaseIntroducer);
+  } else if (context.PositionIs(Lex::TokenKind::Default)) {
+    context.PushState(State::MatchCaseLoopAfterDefault);
+    context.PushState(State::MatchDefaultIntroducer);
+  }
+}
+
+auto HandleMatchCaseLoopAfterDefault(Context& context) -> void {
+  context.PopAndDiscardState();
+
+  std::optional<std::string> kind;
+  if (context.PositionIs(Lex::TokenKind::Case)) {
+    kind = "case";
+  } else if (context.PositionIs(Lex::TokenKind::Default)) {
+    kind = "default";
+  }
+  if (kind) {
```

Comment:
I can see how this handles extra `case` and `default`, but what about other tokens? Some candidate test cases maybe:

```carbon
fn F(x: i32) -> i32 {
  match (x) {
    default => { break; }

    // Bad token `break`
    break;
  }
  match (x) {
    default = { break; }

    // Bad symbols
    { break; }
  }
  return 42;
}
```